### PR TITLE
[DRAFT] Nvidia Settings API changes

### PR DESCRIPTION
**Pull request by monirul**
**Issue number:**

Closes #

**Description of changes:**
This PR introduces new settings API for Nvidia GPUs for Kubernetes Nvidia variants. 

New settings are

Bottlerocket Settings | Impact | Value 
-- | -- | -- 
`settings.nvidia-container-runtime.visible-devices-as-volume-mounts` | allows to change the  `accept-nvidia-visible-devices-as-volume-mounts` value for k8s container-toolkit | `true` \| `false` default: `true`
`settings.nvidia-container-runtime.visible-devices-envvar-when-unprivileged` | allows to set value of  `accept-nvidia-visible-devices-envvar-when-unprivileged` settings of nvidia container runtime for k8s varient | `true` \| `false` default: `false` 
`settings.kubernetes.device-plugins.nvidia.pass-device-specs` | sets the value of the `pass-device-specs` settings of the device plugin that pass the list of DeviceSpecs to the kubelet on Allocate | `true` \| `false` default: `true`
`settings.kubernetes.device-plugins.nvidia.device-id-strategy` | sets the value of the `device-id-strategy` settings of the device plugin which specifies how GPUs are identified and selected for workloads running in a Kubernetes cluster | `uuid` \| `index` Default: `index` 
`settings.kubernetes.device-plugins.nvidia.device-list-strategy` | sets the value of  `device-list-strategy` setting in NVIDIA Kubernetes device plugins. It is used to configure how GPUs are listed and allocated to pods in a Kubernetes cluster | `envvar` \| `volume-mounts` default: `volume-mounts`

**Testing done:**
Yes.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: Release.toml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-version = "1.21.0"
+version = "1.21.1"
```
Comment (bcressey) [inline Release.toml]: Typically we would not add new settings to the API in a point release. These changes should target 1.22. But don't do the release version bump in a feature PR, it's not really related to your feature and creates some churn.

File: Twoliter.toml [hunk 2]
```diff
@@ -1,15 +1,21 @@
 schema-version = 1
-release-version = "1.21.0"
+release-version = "1.21.1"
 
 [vendor.bottlerocket]
 registry = "public.ecr.aws/bottlerocket"
 
+# ======= REVERT BEFORE FINAL PR =================
+[vendor.monirul]
+registry = "397850904080.dkr.ecr.us-west-2.amazonaws.com"
```
Comment (bcressey) [inline Twoliter.toml]: Please don't use non-Bottlerocket vendors in a PR unless it's marked as draft. Too easy for it to be merged by accident.

File: sources/shared-defaults/nvidia-k8s-device-plugin.toml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# nvidia device plugin service
+[services.nvidia-k8s-device-plugin]
+restart-commands = ["/bin/systemctl try-reload-or-restart nvidia-k8s-device-plugin.service"]
+configuration-files = [
+    "nvidia-k8s-device-plugin-conf",
+    "nvidia-k8s-device-plugin-exec-start-conf"
+    ]
```
Comment (bcressey) [inline sources/shared-defaults/nvidia-k8s-device-plugin.toml]: TOML style:
```suggestion
configuration-files = [
    "nvidia-k8s-device-plugin-conf",
    "nvidia-k8s-device-plugin-exec-start-conf",
]
```

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+use migration_helpers::common_migrations::AddPrefixesMigration;
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added a new setting for configuring container runtime (containerd) settings only for NVIDIA k8s variants.
```
Comment (bcressey) [inline sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs]: What's the relationship of the device plugin settings to containerd?

```suggestion
/// We added a new setting for configuring the NVIDIA k8s device plugin.
```

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-metadata/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+use migration_helpers::common_migrations::{AddMetadataMigration, SettingMetadata};
+use migration_helpers::migrate;
+use migration_helpers::Result;
+use std::process;
+
+/// We added a new setting for configuring container runtime (containerd) settings only for NVIDIA k8s variants.
```
Comment (bcressey) [inline sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-metadata/src/main.rs]: Same question about containerd for this comment.

File: sources/settings-migrations/v1.21.1/nvidia-container-runtime-settings/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+use migration_helpers::common_migrations::AddPrefixesMigration;
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added a new setting for configuring container runtime (containerd) settings only for NVIDIA k8s variants.
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec![
+        "settings.nvidia-container-runtime",
```
Comment (bcressey) [inline sources/settings-migrations/v1.21.1/nvidia-container-runtime-settings/src/main.rs:8]:
```suggestion
        "settings.nvidia-container-runtime",
        "services.nvidia-container-toolkit",
        "configuration-files.nvidia-container-toolkit",
```

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+fn run() -> Result<()> {
+    migrate(AddPrefixesMigration(vec![
+        "settings.kubernetes.device-plugins",
```
Comment (bcressey) [inline sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs:8]:
```suggestion
        "settings.kubernetes.device-plugins",
        "services.nvidia-k8s-device-plugin",
        "configuration-files.nvidia-k8s-device-plugin-conf",
        "configuration-files.nvidia-k8s-device-plugin-exec-start-conf",
```

File: packages/settings-plugins/settings-plugins.spec [hunk 8]
```diff
@@ -139,6 +151,11 @@ Conflicts: %{_cross_os}settings-plugin(any)
   -p settings-plugin-vmware-k8s \
   %{nil}
 
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+  -p settings-plugin-aws-k8s-nvidia \
+  %{nil}
```
Comment (bcressey) [inline packages/settings-plugins/settings-plugins.spec:162]: Would be good to add a comment as to why this is built separately (because it needs a different feature flag.)

Can you demonstrate that `settings-plugin-aws-k8s` is built without the `nvidia-device-plugin` feature?

File: sources/settings-plugins/aws-k8s-nvidia/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+git = "https://github.com/bottlerocket-os/bottlerocket-settings-sdk"
+rev = "4709409481ebc0c2f90bb993c257cd9d670190bf"
+version = "0.1.0"
+
+[dependencies.bottlerocket-settings-models]
+# ======= REMOVE BEFORE FINAL PR =================
+git = "https://github.com/bottlerocket-os/bottlerocket-settings-sdk"
+rev = "4709409481ebc0c2f90bb993c257cd9d670190bf"
+version = "0.2.0"
+features = ["nvidia-device-plugin"]
```
Comment (bcressey) [inline sources/settings-plugins/aws-k8s-nvidia/Cargo.toml]: Why define these as workspace dependencies if we're not going to use them?
```suggestion
bottlerocket-settings-plugin = { workspace = true }
bottlerocket-settings-models = { workspace = true, features = ["nvidia-device-plugin"] }
```

File: Twoliter.toml (hunk 2 above)
Comment (monirul) [reply Twoliter.toml]: I marked the PR as draft.

File: packages/settings-plugins/settings-plugins.spec (hunk 8 above)
Comment (monirul) [reply packages/settings-plugins/settings-plugins.spec:162]: I did not understand what do you mean by "Can you demonstrate that settings-plugin-aws-k8s is built without the nvidia-device-plugin feature?" Can you pls elaborate?
The feature flag is provided in the cargo file of that package.

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-metadata/src/main.rs (hunk 5 above)
Comment (monirul) [reply sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-metadata/src/main.rs]: Ahh. Copy paste error. Copied the migration file from the container-runtime and forgot to change the comments.

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs (hunk 4 above)
Comment (monirul) [reply sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs]: Ahh. Copy paste error. Copied the migration file from the container-runtime and forgot to change the comments.

File: packages/settings-plugins/settings-plugins.spec (hunk 8 above)
Comment (bcressey) [reply packages/settings-plugins/settings-plugins.spec:162]: You have two settings plugins:
* `settings-plugin-aws-k8s` - should be built without the feature flag
* `settings-plugin-aws-k8s-nvidia` - should be built with the feature flag

The risk is that with [feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification) you end up with this behavior:
> When a dependency is used by multiple packages, Cargo will use the union of all features enabled on that dependency when building it

So it's not obvious that `settings-plugin-aws-k8s` would be built without the feature flag and you should take steps to confirm that is actually the case.

File: sources/shared-defaults/nvidia-k8s-device-plugin.toml (hunk 3 above)
Comment (monirul) [reply sources/shared-defaults/nvidia-k8s-device-plugin.toml]: changed as suggested.

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs (hunk 4 above)
Comment (monirul) [reply sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs]: changed as suggested.

File: sources/settings-plugins/aws-k8s-nvidia/Cargo.toml (hunk 9 above)
Comment (monirul) [reply sources/settings-plugins/aws-k8s-nvidia/Cargo.toml]: changed as suggested.

File: packages/settings-plugins/settings-plugins.spec (hunk 8 above)
Comment (monirul) [reply packages/settings-plugins/settings-plugins.spec:162]: I have changed the build script to build settings-plugin-aws-k8sand settings-plugin-aws-k8s-nvidia separately to avoid feature flag unification. Also added a comment for that.

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-metadata/src/main.rs (hunk 5 above)
Comment (monirul) [reply sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-metadata/src/main.rs]: changed.

File: sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs (hunk 7 above)
Comment (monirul) [reply sources/settings-migrations/v1.21.1/nvidia-k8s-device-plugin-settings/src/main.rs:8]: changed a suggested.

File: Release.toml [hunk 10]
```diff
@@ -330,3 +330,9 @@ version = "1.21.0"
     "migrate_v1.21.0_k8s-reserved-cpus-v0-1-0.lz4",
     "migrate_v1.21.0_add-hostname-override-source.lz4",
 ]
+"(1.21.0, 1.21.1)" = [
```
Comment (bcressey) [inline Release.toml]:
```suggestion
"(1.21.0, 1.22.0)" = [
```

File: packages/settings-plugins/settings-plugins.spec [hunk 11]
```diff
@@ -132,13 +144,25 @@ Conflicts: %{_cross_os}settings-plugin(any)
...
+# https://doc.rust-lang.org/cargo/reference/features.html#feature-unification
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+  -p settings-plugin-aws-k8s \
+  %{nil}
+
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+  -p settings-plugin-aws-k8s-nvidia \
+  %{nil}
+
+
```
Comment (bcressey) [inline packages/settings-plugins/settings-plugins.spec]: remove one of the two newlines:
```suggestion
```

File: packages/settings-plugins/settings-plugins.spec [hunk 12]
```diff
@@ -57,26 +57,38 @@ Summary: Settings plugin for the aws-k8s variants
...
+Provides: %{_cross_os}settings-plugin(aws-k8s-1.26-nvidia)
+Provides: %{_cross_os}settings-plugin(aws-k8s-1.27-nvidia)
+Provides: %{_cross_os}settings-plugin(aws-k8s-1.28-nvidia)
+Provides: %{_cross_os}settings-plugin(aws-k8s-1.29-nvidia)
+Provides: %{_cross_os}settings-plugin(aws-k8s-1.30-nvidia)
+Conflicts: %{_cross_os}settings-plugin(any)
+
+%description aws-k8s-nvidia
+%{summary}.
+
```
Comment (bcressey) [inline packages/settings-plugins/settings-plugins.spec:90]: avoid adding unnecessary whitespace:
```suggestion
```

**Review by bcressey — COMMENTED** [review]
Looks good apart from the parts that need to be reverted or cleaned up.

It'd be good to test a non-nvidia `aws-k8s` variant to confirm that the device plugin settings aren't recognized, which would indicate that the feature flag wasn't used at build time.

**Comment by cbgbt** [comment]
I spoke with @monirul yesterday about an idea to programmatically verify that feature unification has not taken place. Since we need to do a settings-sdk release for the new models anyways, I think it would be a good idea to make the requisite changes their too.

The basic idea is:
* Add conditionally-compiled `const` booleans for the enabled feature
* Statically assert in the settings models that those flags are as-expected.

**Comment by arnaldo2792** [comment]
This was superseded by #4182.
